DESIGN.md

```diff
@@ -0,0 +1,279 @@
+this host. The `--server` flag is required by miniclient to set this server.
+
+In addition, a Fleetspeak client needs a public key that it will use to check
+the signature of externally provided service configurations. The implications of
+this discussed in more detail below. The `--deployment_key_file` flag is
+required by miniclient to configure this.
+
+To build a binary that can only be used by a specific Fleetspeak installation,
+we recommend creating binary which hardcodes the configuration parameters set by
+`--trusted_cert_file`, `--server`, and `--deployment_key_file`. Specifically, we
+strongly recommend this if you will sign, whitelist a hash, or otherwise mark
+the Fleetspeak client binary as trusted.
+
+## Configuration Path
+
+In addition to the security critical parameters described in the previous
+section, a Fleetspeak client normally requires a configuration directory to
+store its private key and to look for additional configuration. See the comments
+on [`config.Configuration`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/config#Configuration) for details.
+
+## Client Communicator
+
+Every client requires a
+[`Communicator`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/comms#Communicator)
+component which handles communication with the server. This component defines how
+the client communicates with the Fleetspeak server and should agree on protocol
+with a corresponding [Server Communicator](#server-communicator).
+
+The quintessential example is
+[`https.Communicator`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/https#Communicator)
+and at this point it is the expected implementation for essentially all
+installations. By adjusting the `DialContext` attribute before starting the
+Fleetspeak client, it should be possible to include any needed additional
+support for specialized tunneling or proxying.
+
+## Client Service Factories and Configuration
+
+When messages arrive on a Fleetspeak client, they must be addressed to a
+particular service - the `destination` field as described in
+[Messages](#messages).  A Fleetspeak client runs one or more Services in order
+to handle these messages, and create message to send to the server.  By
+configuring additional independent services a single Fleetspeak client process
+can handle messages for independent purposes.
+
+A service is typically configured by dropping a
+[`fleetspeak.SignedClientServiceConfig`](https://github.com/google/fleetspeak/blob/master/fleetspeak/src/common/proto/fleetspeak/system.proto)
+into `<config_path>/services/`. Besides the name, used to address the service,
+the most important parameter is `factory`. This string is used to look up a
+[`service.Factory`](https://godoc.org/github.com/google/fleetspeak/fleetspeak/src/client/service#Factory),
+which determines what code to use to process incoming messages.
+
+### Stdinservice
+
+The
```

Comment:
I find this to be a complex explanation of a simple concept. We should just say something like "runs a given command line with given input and returns the output"

Reply:
Tried to simplify.

Reply:
Better. :+1: